build-system/tasks/performance-test/measure-documents.js

```diff
@@ -0,0 +1,196 @@
+    );
+
+    layoutShiftObserver.observe({entryTypes: ['layout-shift']});
+
+    const largestContentfulPaintObserver = new PerformanceObserver(list => {
+      const entries = list.getEntries();
+      const entry = entries[entries.length - 1];
+      window.largestContentfulPaint = entry.renderTime || entry.loadTime;
+    });
+
+    largestContentfulPaintObserver.observe({
+      entryTypes: ['largest-contentful-paint'],
+    });
+  });
+
+/**
+ * Evaluate script on the page to collect and calculate metrics
+ *
+ * @param {Puppeteer.page} page
+ * @return {Promise<object>} Resolves with page load metrics
+ */
+const readMetrics = page =>
+  page.evaluate(() => {
+    const entries = performance.getEntries();
+
+    function getMetric(name) {
+      const entry = entries.find(entry => entry.name === name);
+      return entry ? Math.round(entry.startTime) : 0;
+    }
+
+    const firstPaint = getMetric('first-paint');
+    const firstContentfulPaint = getMetric('first-contentful-paint');
+
+    function getMaxFirstInputDelay() {
+      let longest = 0;
+
+      window.longTasks.forEach(longTask => {
+        if (
+          longTask.startTime > firstContentfulPaint &&
+          longTask.duration > longest
+        ) {
+          longest = longTask.duration;
+        }
+      });
+
+      return longest;
+    }
+
+    function getTimeToInteractive() {
+      return Date.now() - window.measureStarted;
+    }
+
+    return {
+      visible: getMetric('visible'),
+      firstPaint,
+      firstContentfulPaint,
+      largestContentfulPaint: window.largestContentfulPaint,
+      timeToInteractive: getTimeToInteractive(),
+      maxFirstInputDelay: getMaxFirstInputDelay(),
+      cumulativeLayoutShift: window.cumulativeLayoutShift,
+    };
+  });
+
+/**
+ * Writes measurements to ./results.json
+ *
+ * @param {string} version
+ * @param {*} metrics
+ */
+function writeMetrics(version, metrics) {
+  let results = {
+    [CONTROL]: [],
+    [EXPERIMENT]: [],
+  };
+
+  if (fs.existsSync(RESULTS_PATH)) {
+    results = JSON.parse(fs.readFileSync(RESULTS_PATH));
+  }
+
+  results[version] = [...results[version], metrics];
+  fs.writeFileSync(RESULTS_PATH, JSON.stringify(results));
+}
+
+/**
+ * Opens Chrome, loads the URL from local file cache, and collects
+ * metrics for the specified URL and version
+ *
+ * @param {string} url
+ * @param {string} version "control" or "experiment"
+ * @param {Object} options
+ * @return {Promise}
+ */
+async function measureDocument(url, version, {headless}) {
+  const browser = await puppeteer.launch({
+    headless,
+    args: [
+      '--allow-file-access-from-files',
+      '--enable-blink-features=LayoutInstabilityAPI',
+    ],
+  });
+
+  const page = await browser.newPage();
+  await page.setCacheEnabled(false);
+  await setupMeasurement(page);
+
+  try {
+    await page.goto(`file:${urlToCachePath(url, version)}`, {
+      waitUntil: 'networkidle0',
```

Comment:
FYI we had some issues with `networkidle0`, sometimes websites keep an open connection (iframed ads tend to do that) for longer than the default timeout. `networkidle2` worked as a compromise for us with visual diffs. If you get lots of errors here, try that.

Might be worth adding logging statements in the catch block to see how often that happens